tools/c7n_oci/c7n_oci/resources/compute.py

```diff
@@ -0,0 +1,288 @@
+
+import logging
+import re  # noqa
+import copy  # noqa
+
+import oci.core
+
+from c7n.filters import Filter, ValueFilter  # noqa
+from c7n.utils import type_schema
+from c7n_oci.actions.base import OCIBaseAction, RemoveTagBaseAction
+from c7n_oci.provider import resources
+from c7n_oci.query import QueryResourceManager
+
+log = logging.getLogger("custodian.oci.resources.compute")
+
+
+@resources.register("instance")
+class Instance(QueryResourceManager):
+    """Oracle Cloud Infrastructure Instance Resource
+
+    :example:
+
+    Returns all Instance resources in the tenancy
+
+    .. code-block:: yaml
+
+        policies:
+            - name: find-all-instance-resources
+              resource: oci.instance
+
+    """
+
+    class resource_type:
+        doc_groups = ["Compute"]
+        service = "oci.core"
+        client = "ComputeClient"
+        enum_spec = ("list_instances", "items[]", None)
+        extra_params = {"compartment_id", "instance_id"}
+        resource_type = "OCI.Compute/Instance"
+        id = "id"
+        name = "display_name"
+        search_resource_type = "instance"
+
+
+@Instance.filter_registry.register("monitoring")
+class InstanceMonitoring(Filter):
+    """
+    Instance Monitoring Filter
+
+    :example:
+
+    This filter returns the resources with the aggregated data that match the criteria specified in the request.
+    Compartment OCID required. For information on metric queries, see `Building Metric Queries
+    <https://docs.oracle.com/en-us/iaas/Content/Monitoring/Tasks/buildingqueries.htm>`_.
+
+    .. code-block:: yaml
+
+        policies:
+            - name: instance-with-low-cpu-utilization
+            description: |
+                Return the instances with the low CPU utilization
+            resource: oci.instance
+            filters:
+                - type: monitoring
+                  query: 'CpuUtilization[30d].mean() < 6'
+
+    """  # noqa
+
+    schema = type_schema("monitoring", query={"type": "string"}, required=["query"])
+
+    def process(self, resources, event):
+        summarize_metrics = oci.monitoring.models.SummarizeMetricsDataDetails(
+            query=self.data.get("query"),
+            namespace="oci_computeagent",
+        )
+        monitoring_client = self.manager.get_session().get_monitoring_client()
+        comp_resources = {}
+        result = []
+        for resource in resources:
+            comp_id = resource.get("compartment_id")
+            if comp_id in comp_resources:
+                comp_resources.get(comp_id).append(resource)
+            else:
+                comp_resources[comp_id] = [resource]
+        # Query the MonitoringClient with the query against each compartment and perform
+        # the filtering
+        for compartment_id in comp_resources.keys():
+            metric_response = monitoring_client.summarize_metrics_data(
+                compartment_id=compartment_id,
+                summarize_metrics_data_details=summarize_metrics,
```

Comment:
just thinking through on this. the syntax offered by the service team for query, is pretty cool. but if we've filtered down to a single instance (say event based on change), and the compartment contains hundreds/thousands this feels like we're potentially pulling in extraneous data, versus just the instance[s] of interest. can we pass the instance list for the summarization via query customization, and if so what's the query size limit.

Reply:
This is an optimization which we can look into. It is dependent on what the monitoring service API will support. We would have to look into the feasibility of this and would therefore like to push it to a later release.

Reply:
okay, let's file as a separate GitHub issue then.